### PR TITLE
Clarify that ListOptions.Timeout is not conditional on inactivity

### LINKED ISSUE #58558: metav1.ListOptions.Timeout documentation is unclear
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Watches done with the `Timeout` field of `ListOptions` set to some value T end after T regardless of whether there has been any activity in the meantime.

**What you expected to happen**:
I expected the watch to end when no events have been delivered in the last T.  In ordinary usage, a "timeout" is something that happens _if_ there has been no activity for the specified length of time.

**How to reproduce it (as minimally and precisely as possible)**:
Create a configmap named `b` in namespace `foo`.  Then in one shell thrash it periodically like this:

```
$ while true; do kubectl patch configmap b -n foo -p '{"data": {"b": "x"}}'; sleep 7; kubectl patch configmap b -n foo -p '{"data": {"b": "y"}}'; sleep 7; done
```
Observe that it proceeds to update the configmap every 7 seconds or so.

In another shell do a watch with a timeout; observe that it ends even though the thrashing continues.
```
$ date; curl 'http://localhost:8080/api/v1/namespaces/foo/configmaps?watch=true&timeoutSeconds=42'; date
Sat Jan 20 01:44:31 UTC 2018
{"type":"ADDED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205310","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"y"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205577","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"x"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205591","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"y"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205605","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"x"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205619","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"y"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205634","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"x"}}}
{"type":"MODIFIED","object":{"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"b","namespace":"foo","selfLink":"/api/v1/namespaces/foo/configmaps/b","uid":"2769c179-fd64-11e7-8d7f-525400178201","resourceVersion":"2205647","creationTimestamp":"2018-01-19T22:00:25Z"},"data":{"b":"y"}}}
Sat Jan 20 01:45:13 UTC 2018
$ 
```

**Anything else we need to know?**:

**Environment**:

- Kubernetes version (use `kubectl version`): Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.4+coreos.0", GitCommit:"4292f9682595afddbb4f8b1483673449c74f9619", GitTreeState:"clean", BuildDate:"2017-11-21T17:22:25Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: local hardware
- OS (e.g. from /etc/os-release): Ubuntu 16.04.3 LTS
- Kernel (e.g. `uname -a`): Linux n17s2ctrl 4.4.0-34-generic #53-Ubuntu SMP Wed Jul 27 16:06:39 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: local
- Others:

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What this PR does / why we need it**:
Clarifies the meaning of the `Timeout` field of a `ListOptions`, because the name and existing description can reasonably be interpreted to mean something divergent from the actual behavior.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58558

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
CC @lavalamp 
CC @liggitt

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -342,6 +342,8 @@ type ListOptions struct {
 	// +optional
 	ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,4,opt,name=resourceVersion"`
 	// Timeout for the list/watch call.
+	// For a watch call, this is a limit on the duration of the call,
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: suggest "This limits the duration of the call, regardless of any activity or inactivity."

**Comment by liggitt** [comment]
see failures in verify scripts... generated docs need updating

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Thanks for the suggestion.  Does that apply to `list` as well as `watch`?

**Comment by MikeSpreitzer** [comment]
/lgtm cancel

with a heavy heart, but something is going wrong with that verify job.

**Comment by MikeSpreitzer** [comment]
@sttts , @liggitt , @hongchaodeng : I have rebased and re-generated again.  Ready for another LGTM.

**Pull request merged**